pulsar-io/azure-data-explorer/src/main/java/org/apache/pulsar/io/azuredataexplorer/ADXSink.java

```diff
@@ -0,0 +1,184 @@
+
+    IngestClient client;
+    IngestionProperties ingestionProperties;
+    private static final Logger LOG = LoggerFactory.getLogger(ADXSink.class);
+    private List<Record<byte[]>> incomingRecordsList;
+    private int batchSize;
+    private ScheduledExecutorService adxSinkExecutor;
+
+    ObjectMapper mapper;
+
+
+    @Override
+    public void open(Map<String, Object> config, SinkContext sinkContext) throws Exception {
+
+        // Azure data explorer, initializations
+        ADXSinkConfig adxconfig = ADXSinkConfig.load(config);
+        ConnectionStringBuilder kcsb = getConnectionStringBuilder(adxconfig, getAuthenticationMode(adxconfig));
+        if (kcsb == null) throw new Exception("Kusto Connection String NULL");
+        LOG.debug("ConnectionString created:  " + kcsb);
+        client = adxconfig.isManagedIngestion() ? IngestClientFactory.createManagedStreamingIngestClient(kcsb) : IngestClientFactory.createClient(kcsb);
+        ingestionProperties = new IngestionProperties(adxconfig.getDatabase(),
+                adxconfig.getTable());
+        ingestionProperties.setIngestionMapping(adxconfig.getMappingRefName(), getParseMappingRefType(adxconfig.getMappingRefType()));
+        ingestionProperties.setFlushImmediately(adxconfig.isFlushImmediately());
+        ingestionProperties.setDataFormat(IngestionProperties.DataFormat.JSON);
+        LOG.debug("Ingestion Properties:  " + ingestionProperties.toString());
+
+        //incoming records list will hold incoming messages, flushExecutor executes the flushData according to batch time
+        batchSize = adxconfig.getBatchSize();
+        long batchTimeMs = adxconfig.getBatchTimeMs();
+        incomingRecordsList = new ArrayList<>();
+        adxSinkExecutor = Executors.newScheduledThreadPool(1);
+        adxSinkExecutor.scheduleAtFixedRate(this::sinkData, batchTimeMs, batchTimeMs, TimeUnit.MILLISECONDS);
+        mapper = new ObjectMapper();
+
+    }
+
+    @Override
+    public void write(Record<byte[]> record) throws Exception {
+        int runningSize = 0;
+        synchronized (this) {
+            incomingRecordsList.add(record);
+            runningSize = incomingRecordsList.size();
+        }
+        if (runningSize == batchSize) {
+            adxSinkExecutor.execute(this::sinkData);
+        }
+    }
+
+    private void sinkData() {
+
+        final List<Record<byte[]>> recordsToSink;
+
+        synchronized (this) {
+            if (incomingRecordsList.isEmpty()) return;
+            recordsToSink = incomingRecordsList;
+            incomingRecordsList = new ArrayList<>();
+        }
+
+        for (Record<byte[]> record : recordsToSink) {
+            try {
+                ADXPulsarEvent event = getADXPulsarEvent(record);
+                StreamSourceInfo streamSourceInfo = new StreamSourceInfo(new ByteArrayInputStream(mapper.writeValueAsString(event).getBytes()));
+                OperationStatus status = client.ingestFromStream(streamSourceInfo,
```

Comment:
Status checking is tricky! We attempted to avoid the "idling" (Thread.sleep) in Flink using a thread logic.
A fanstastic thread [here](https://stackoverflow.com/questions/40251528/how-to-use-executorservice-to-poll-until-a-result-arrives)